### ISSUE TITLE
Recursive Comparison With Ignored Fields Fails On Equal Collections

### ISSUE THREAD

**Issue by nicoprow** [assertj/assertj#3512]
**Describe the bug**
When recursively comparing two lists that are equal, the last expected element is not found in the actual list of elements. We observed this bug when ignoring collection order, ignoring three fields and two datatypes.

In contrast to the existing bug #3226 all the ignored fields exists in the expected list.

Remark: We are using Kotlin.

* assertj core version: 3.26.0
* java version: 21.0.2
* kotlin version: 1.9.24

We observed this bug when upgrading from assertj 3.24.2 to 3.26.0. (No changes to tests or code have been done for this upgrade.)

**Test case reproducing the bug**

Sadly, I could not reproduce the bug on a smaller dataset. So I will share the bug how it occurs in our test.

```bash
Top level actual and expected objects differ:
- actual value  : [LogisticAddressVerboseDto(bpna=BPNA0000000002XY, name=null, states=[], identifiers=[], physicalPostalAddress=PhysicalPostalAddressVerboseDto(geographicCoordinates=null, countryVerbose=TypeKeyNameVerboseDto(technicalKey=DE, name=Germany), administrativeAreaLevel1Verbose=null, administrativeAreaLevel2=Böblingen, administrativeAreaLevel3=null, postalCode=71059 , city=Böblingen, district=Sindelfingen-Ost, street=StreetDto(name=Bela-Barenyi-Straße, houseNumber=1, houseNumberSupplement=A, milestone=milestone 1, direction=direction 1, namePrefix=name prefix 1, additionalNamePrefix=add name prefix 1, nameSuffix=name suffix 1, additionalNameSuffix=add name suffix 1), companyPostalCode=70547, industrialZone=Industrial Zone One, building=Gebäude eins, floor=Stockerk eins, door=Raum eins), alternativePostalAddress=null, bpnLegalEntity=null, bpnSite=BPNS000000000001, isCatenaXMemberData=false, createdAt=2024-06-19T04:01:31.259451Z, updatedAt=2024-06-19T04:01:31.264620Z, confidenceCriteria=ConfidenceCriteriaDto(sharedByOwner=true, checkedByExternalDataSource=true, numberOfSharingMembers=1, lastConfidenceCheckAt=2023-10-10T10:10:10, nextConfidenceCheckAt=2024-10-10T10:10:10, confidenceLevel=10), addressType=SiteMainAddress),
    LogisticAddressVerboseDto(bpna=BPNA0000000004WK, name=null, states=[], identifiers=[], physicalPostalAddress=PhysicalPostalAddressVerboseDto(geographicCoordinates=null, countryVerbose=TypeKeyNameVerboseDto(technicalKey=DE, name=Germany), administrativeAreaLevel1Verbose=null, administrativeAreaLevel2=Böblingen, administrativeAreaLevel3=null, postalCode=71059 , city=Böblingen, district=Sindelfingen-Ost, street=StreetDto(name=Bela-Barenyi-Straße, houseNumber=1, houseNumberSupplement=A, milestone=milestone 1, direction=direction 1, namePrefix=name prefix 1, additionalNamePrefix=add name prefix 1, nameSuffix=name suffix 1, additionalNameSuffix=add name suffix 1), companyPostalCode=70547, industrialZone=Industrial Zone One, building=Gebäude eins, floor=Stockerk eins, door=Raum eins), alternativePostalAddress=null, bpnLegalEntity=null, bpnSite=BPNS000000000001, isCatenaXMemberData=false, createdAt=2024-06-19T04:01:31.346893Z, updatedAt=2024-06-19T04:01:31.346900Z, confidenceCriteria=ConfidenceCriteriaDto(sharedByOwner=true, checkedByExternalDataSource=true, numberOfSharingMembers=1, lastConfidenceCheckAt=2023-10-10T10:10:10, nextConfidenceCheckAt=2024-10-10T10:10:10, confidenceLevel=10), addressType=AdditionalAddress),
    LogisticAddressVerboseDto(bpna=BPNA0000000005VV, name=null, states=[], identifiers=[], physicalPostalAddress=PhysicalPostalAddressVerboseDto(geographicCoordinates=null, countryVerbose=TypeKeyNameVerboseDto(technicalKey=FR, name=France), administrativeAreaLevel1Verbose=null, administrativeAreaLevel2= Fulton County, administrativeAreaLevel3=null, postalCode=70547, city=Atlanta, district=District Level 1, street=StreetDto(name=Street2, houseNumber=2, houseNumberSupplement=B, milestone=milestone 2, direction=direction 2, namePrefix=name prefix 2, additionalNamePrefix=add name prefix 2, nameSuffix=name suffix 2, additionalNameSuffix=add name suffix 2), companyPostalCode=null, industrialZone=Industrial Zone Two, building=Building Two, floor=Floor Two, door=Door Two), alternativePostalAddress=null, bpnLegalEntity=null, bpnSite=BPNS000000000001, isCatenaXMemberData=false, createdAt=2024-06-19T04:01:31.347222Z, updatedAt=2024-06-19T04:01:31.347225Z, confidenceCriteria=ConfidenceCriteriaDto(sharedByOwner=false, checkedByExternalDataSource=false, numberOfSharingMembers=3, lastConfidenceCheckAt=2022-10-10T10:10:10, nextConfidenceCheckAt=2025-10-10T10:10:10, confidenceLevel=6), addressType=AdditionalAddress)]
- expected value: [LogisticAddressVerboseDto(bpna=BPNA000000000001, name=null, states=[], identifiers=[], physicalPostalAddress=PhysicalPostalAddressVerboseDto(geographicCoordinates=null, countryVerbose=TypeKeyNameVerboseDto(technicalKey=DE, name=Germany), administrativeAreaLevel1Verbose=null, administrativeAreaLevel2=Böblingen, administrativeAreaLevel3=null, postalCode=71059 , city=Böblingen, district=Sindelfingen-Ost, street=StreetDto(name=Bela-Barenyi-Straße, houseNumber=1, houseNumberSupplement=A, milestone=milestone 1, direction=direction 1, namePrefix=name prefix 1, additionalNamePrefix=add name prefix 1, nameSuffix=name suffix 1, additionalNameSuffix=add name suffix 1), companyPostalCode=70547, industrialZone=Industrial Zone One, building=Gebäude eins, floor=Stockerk eins, door=Raum eins), alternativePostalAddress=null, bpnLegalEntity=null, bpnSite=null, isCatenaXMemberData=false, createdAt=2024-06-19T04:01:30.454922Z, updatedAt=2024-06-19T04:01:30.454924Z, confidenceCriteria=ConfidenceCriteriaDto(sharedByOwner=true, checkedByExternalDataSource=true, numberOfSharingMembers=1, lastConfidenceCheckAt=2023-10-10T10:10:10, nextConfidenceCheckAt=2024-10-10T10:10:10, confidenceLevel=10), addressType=SiteMainAddress),
    LogisticAddressVerboseDto(bpna=BPNA000000000001, name=null, states=[], identifiers=[], physicalPostalAddress=PhysicalPostalAddressVerboseDto(geographicCoordinates=null, countryVerbose=TypeKeyNameVerboseDto(technicalKey=DE, name=Germany), administrativeAreaLevel1Verbose=null, administrativeAreaLevel2=Böblingen, administrativeAreaLevel3=null, postalCode=71059 , city=Böblingen, district=Sindelfingen-Ost, street=StreetDto(name=Bela-Barenyi-Straße, houseNumber=1, houseNumberSupplement=A, milestone=milestone 1, direction=direction 1, namePrefix=name prefix 1, additionalNamePrefix=add name prefix 1, nameSuffix=name suffix 1, additionalNameSuffix=add name suffix 1), companyPostalCode=70547, industrialZone=Industrial Zone One, building=Gebäude eins, floor=Stockerk eins, door=Raum eins), alternativePostalAddress=null, bpnLegalEntity=null, bpnSite=null, isCatenaXMemberData=false, createdAt=2024-06-19T04:01:30.454922Z, updatedAt=2024-06-19T04:01:30.454924Z, confidenceCriteria=ConfidenceCriteriaDto(sharedByOwner=true, checkedByExternalDataSource=true, numberOfSharingMembers=1, lastConfidenceCheckAt=2023-10-10T10:10:10, nextConfidenceCheckAt=2024-10-10T10:10:10, confidenceLevel=10), addressType=AdditionalAddress),
    LogisticAddressVerboseDto(bpna=BPNA0000000001YN, name=null, states=[], identifiers=[], physicalPostalAddress=PhysicalPostalAddressVerboseDto(geographicCoordinates=null, countryVerbose=TypeKeyNameVerboseDto(technicalKey=FR, name=France), administrativeAreaLevel1Verbose=null, administrativeAreaLevel2= Fulton County, administrativeAreaLevel3=null, postalCode=70547, city=Atlanta, district=District Level 1, street=StreetDto(name=Street2, houseNumber=2, houseNumberSupplement=B, milestone=milestone 2, direction=direction 2, namePrefix=name prefix 2, additionalNamePrefix=add name prefix 2, nameSuffix=name suffix 2, additionalNameSuffix=add name suffix 2), companyPostalCode=null, industrialZone=Industrial Zone Two, building=Building Two, floor=Floor Two, door=Door Two), alternativePostalAddress=null, bpnLegalEntity=null, bpnSite=null, isCatenaXMemberData=false, createdAt=2024-06-19T04:01:30.455664Z, updatedAt=2024-06-19T04:01:30.455664Z, confidenceCriteria=ConfidenceCriteriaDto(sharedByOwner=false, checkedByExternalDataSource=false, numberOfSharingMembers=3, lastConfidenceCheckAt=2022-10-10T10:10:10, nextConfidenceCheckAt=2025-10-10T10:10:10, confidenceLevel=6), addressType=AdditionalAddress)]
The following expected elements were not matched in the actual ArrayList:
  [LogisticAddressVerboseDto(bpna=BPNA0000000001YN, name=null, states=[], identifiers=[], physicalPostalAddress=PhysicalPostalAddressVerboseDto(geographicCoordinates=null, countryVerbose=TypeKeyNameVerboseDto(technicalKey=FR, name=France), administrativeAreaLevel1Verbose=null, administrativeAreaLevel2= Fulton County, administrativeAreaLevel3=null, postalCode=70547, city=Atlanta, district=District Level 1, street=StreetDto(name=Street2, houseNumber=2, houseNumberSupplement=B, milestone=milestone 2, direction=direction 2, namePrefix=name prefix 2, additionalNamePrefix=add name prefix 2, nameSuffix=name suffix 2, additionalNameSuffix=add name suffix 2), companyPostalCode=null, industrialZone=Industrial Zone Two, building=Building Two, floor=Floor Two, door=Door Two), alternativePostalAddress=null, bpnLegalEntity=null, bpnSite=null, isCatenaXMemberData=false, createdAt=2024-06-19T04:01:30.455664Z, updatedAt=2024-06-19T04:01:30.455664Z, confidenceCriteria=ConfidenceCriteriaDto(sharedByOwner=false, checkedByExternalDataSource=false, numberOfSharingMembers=3, lastConfidenceCheckAt=2022-10-10T10:10:10, nextConfidenceCheckAt=2025-10-10T10:10:10, confidenceLevel=6), addressType=AdditionalAddress)]

The recursive comparison was performed with this configuration:
- the following fields were ignored in the comparison: bpna, bpnLegalEntity, bpnSite
- the following types were ignored in the comparison: java.time.Instant, java.time.LocalDateTime
- no equals methods were used in the comparison EXCEPT for java JDK types since introspecting JDK types is forbidden in java 17+ (use withEqualsForType to register a specific way to compare a JDK type if you need it)
- collection order was ignored in all fields in the comparison
- these types were compared with the following comparators:
  - java.lang.Double -> DoubleComparator[precision=1.0E-15]
  - java.lang.Float -> FloatComparator[precision=1.0E-6]
  - java.nio.file.Path -> lexicographic comparator (Path natural order)
- actual and expected objects and their fields were compared field by field recursively even if they were not of the same type, this allows for example to compare a Person to a PersonDto (call strictTypeChecking(true) to change that behavior).
- the introspection strategy used was: DefaultRecursiveComparisonIntrospectionStrategy
```

Comparison Code:
```kotlin
assertThat(actuals).usingRecursiveComparison()
            .ignoringFields(
                LogisticAddressVerboseDto::bpna.name,
                LogisticAddressVerboseDto::bpnLegalEntity.name,
                LogisticAddressVerboseDto::bpnSite.name,
            )
            .ignoringCollectionOrder()
            .ignoringFieldsOfTypes(Instant::class.java, LocalDateTime::class.java)
            .isEqualTo(expected)
```

Diff-Check Comparison for the element that is said to be missing:
<img width="1088" alt="grafik" src="https://github.com/assertj/assertj/assets/95762387/9160e1a8-7af3-4b61-882d-7818fd3721e6">

As you can see the only fields changed here are "bpna", "bpnSite" and the "createdAt" and "updatedAt" timestamps which are ignored due to ignoring the Instant field type.

The bug itself seems to be quite fickle as the comparison succeeds when changing certain comparison options. However, we could not reliably reproduce which configuration leads to the bug actually. For example removing ignoring the collection order now successfully compares the lists as equal: (In that test the elements of two collections are actually in the correct order)

```kotlin
assertThat(actuals).usingRecursiveComparison()
            .ignoringFields(
                LogisticAddressVerboseDto::bpna.name,
                LogisticAddressVerboseDto::bpnLegalEntity.name,
                LogisticAddressVerboseDto::bpnSite.name,
            )
            .ignoringFieldsOfTypes(Instant::class.java, LocalDateTime::class.java)
            .isEqualTo(expected)
```

**Comment by joel-costigliola** [comment]
Thanks for reporting this @nicoprow !

**Comment by joel-costigliola** [comment]
@nicoprow could you share the data structure being compared if not confidential ?

**Comment by joel-costigliola** [comment]
I haven't yet been able to reproduce the issue, I suspect it is related to ignoring collection order as this implies comparing all elements of the collection under test with all elements of the expected one and you said the test worked with an ordered list.

**Comment by scordio** [comment]
@nicoprow we plan to release 3.26.1 in roughly one week.

In case you have time to provide the details mentioned by @joel-costigliola, we might try to squeeze a solution in it 🙂

**Comment by justwrote** [comment]
I have the same issue and cannot upgrade. I tested the SNAPSHOT and it got fixed in https://github.com/assertj/assertj/commit/c5fa7eb9cbba4c726e10908d3e91b9557884dd4b. Would be nice to have a new release.

**Comment by joel-costigliola** [comment]
Let's assume it is indeed fixed in 3.27.0 yet to be released.
